[OpenMP] Migrate GPU Reductions CodeGen from Clang to OMPIRBuilder

TIFitis opened the pull request:
This patch migrates the CGOpenMPRuntimeGPU::emitReduction and related functions to the OpenMPIRBUilder. In future patches MLIR OpenMP translation would be making use of these functions.

Co-authored-by: Jan Leyonberg <jan.leyonberg@amd.com>

TIFitis commented on clang/test/OpenMP/nvptx_teams_reduction_codegen.cpp:
```diff
@@ -2953,7 +2953,7 @@ int bar(int n){
 // CHECK3-NEXT:    [[TMP8:%.*]] = getelementptr inbounds [[STRUCT__GLOBALIZED_LOCALS_TY_0:%.*]], ptr [[TMP4]], i32 [[TMP5]]
 // CHECK3-NEXT:    [[C:%.*]] = getelementptr inbounds [[STRUCT__GLOBALIZED_LOCALS_TY_0]], ptr [[TMP8]], i32 0, i32 0
 // CHECK3-NEXT:    [[TMP9:%.*]] = load i8, ptr [[TMP7]], align 1
-// CHECK3-NEXT:    store i8 [[TMP9]], ptr [[C]], align 4
```
I'm not sure if this change is correct. My intuition says that the alignment was incorrect and this now fixes it, hopefully someone can comment.

jdoerfert replied:
Similar to above, this affects only accesses to a static global. Even if the new alignment is lower than optimal, we should be able to reconstruct it from the allocation. This is fine, IMHO.

TIFitis commented on clang/test/OpenMP/nvptx_target_parallel_reduction_codegen_tbaa_PR46146.cpp:
```diff
@@ -36,14 +36,14 @@ void test() {
 // CHECK1-NEXT:    [[DYN_PTR_ADDR:%.*]] = alloca ptr, align 8
 // CHECK1-NEXT:    [[DOTZERO_ADDR:%.*]] = alloca i32, align 4
 // CHECK1-NEXT:    [[DOTTHREADID_TEMP_:%.*]] = alloca i32, align 4
-// CHECK1-NEXT:    store ptr [[DYN_PTR]], ptr [[DYN_PTR_ADDR]], align 8, !tbaa [[TBAA10:![0-9]+]]
```
From what I've heard from @jsjodin the TBAA info was pending to be removed. There is no good way to have OMPIRBuilder generate the TBAA so it has been muted for reductions codegen, let me know if this change is acceptable.

Otherwise, we would have to look into callbacks or other methods to retain this information.

jdoerfert replied:
Browsing through these, it seems we only loose TBAA for accesses of allocas. This is fine.

TIFitis commented:
@jsjodin This PR only has the migration work. Flang/MLIR is left unchanged other than the few necessary changes.

Once this patch is merged, I'll put up a PR for having MLIR use these additions.

jsjodin commented:
> @jsjodin This PR only has the migration work. Flang/MLIR is left unchanged other than the few necessary changes.
> 
> Once this patch is merged, I'll put up a PR for having MLIR use these additions.

There is some code that is related to Flang/MLIR that shouldn't be in the patch. e.g. ReductionInfoManager is not used anywhere. I am currently working on a patch that refactors this code a bit and how it is used in the MLIR lowering.

TIFitis commented:
> There is some code that is related to Flang/MLIR that shouldn't be in the patch. e.g. ReductionInfoManager is not used anywhere. I am currently working on a patch that refactors this code a bit and how it is used in the MLIR lowering.

Thanks for pointing this out, I've removed it from this PR.

TIFitis commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -2719,10 +4051,24 @@ getKmpcForStaticLoopForType(Type *Ty, OpenMPIRBuilder *OMPBuilder,
   IRBuilder<> &Builder = OMPBuilder->Builder;
+  Value *TripCount = TripCountOrig;
+  // FIXME(JAN): The trip count is 1 larger than it should be for GPU, this may
+  // not be the right way to fix it, but this works for now.
```
This section of the code is never executed from Clang, so I'm not sure.

TIFitis commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -2042,35 +2057,1378 @@ OpenMPIRBuilder::createSection(const LocationDescription &Loc,
+      break;
+    }
+    case EvaluationKindTy::Complex: {
+      // FIXME(Jan): Complex type
```
Yes there are currently no tests for the Complex type, only for scalar and aggregate.

I don't see any other handling of complex type in the OMPIRBuilder, so I'm working on what needs to be done here. In the meanwhile, I've added assertion failures for hitting complex types.

jdoerfert reviewed (commented):
So, just to double check. On the Clang side, all you did is move the codegen and adapt it to work in the OMPIRBuilder, right? Expected outcome is the same, modulo the two small changes.

TIFitis commented:
Yup, code generated is identical, except for the align issue I've highlighted and some cast instructions being moved around. Also no TBAA Info.

TIFitis commented:
With the latest changes the build bots are all clean now. Let me know if there are any further changes required for this patch to go through.

TIFitis commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -2051,36 +2057,1424 @@ OpenMPIRBuilder::createSection(const LocationDescription &Loc,
+        Builder.CreateStore(Elem, DestElementAddr);
+        break;
+      }
+      case EvaluationKindTy::Complex: {
+        assert(false && "Complex data type not handled");
+        break;
+      }
```
@jdoerfert Currently there are no test cases for the Complex kind. I was trying to come up with test cases that might fall under this case but I wasn't able to.

Please let me know if this case is infeasible, and we can leave it as an assertion failure.

Here's my attempt at creating a test case, but it still falls under the aggregate kind:

```
int foo() {
  int i;
  int j;
  std::complex<int> sum[10][10];
  std::complex<int> res;

  #pragma omp target teams loop reduction(+:sum)
  for(i=0; i<10; i++)
    for(j=0; j<10; j++)
      res += sum[i][j];

  return 0;
}
```

TIFitis commented:
Ping for review :)

TIFitis commented:
@jsjodin @jdoerfert I have now added the missing support for complex variable reductions. Please let me know if this patch requires any other changes.

Thanks.

TIFitis commented:
Ping for review.

TIFitis commented:
@alexey-bataev I've added the test clang/test/OpenMP/reduction_complex.c to this PR. I tried to add the checks based on the other similar tests in the directory, let me know if it needs any changes.

alexey-bataev reviewed (commented):
The test looks good

TIFitis commented:
@jdoerfert @jsjodin polite ping for review.

jdoerfert commented on llvm/include/llvm/Frontend/OpenMP/OMPIRBuilder.h:
```diff
@@ -607,15 +621,17 @@ class OpenMPIRBuilder {
   ///
   /// \param Loc The location where the barrier directive was encountered.
-  /// \param DK The kind of directive that caused the barrier.
+  /// \param Kind The kind of directive that caused the barrier.
```
The argument is still called DK.

jdoerfert commented on llvm/include/llvm/Frontend/OpenMP/OMPIRBuilder.h:
```diff
@@ -1235,27 +1251,56 @@ class OpenMPIRBuilder {
-  using ReductionGenTy =
-      function_ref<InsertPointTy(InsertPointTy, Value *, Value *, Value *&)>;
+  /// Enum class for the RedctionGen CallBack type to be used.
+  enum class ReductionGenCBTy { Clang, MLIR };
```
s/Ty/Kind

jdoerfert commented on llvm/include/llvm/Frontend/OpenMP/OMPIRBuilder.h:
```diff
@@ -1235,27 +1251,56 @@ class OpenMPIRBuilder {
+  /// return the RHSPtr it used for codegen, used for fixup later.
+  /// \param CurFn Optionally used by Clang to pass in the Current Function as
+  /// Clang context may be old.
+  using ReductionGenCBClang =
```
s/ReductionGenCBClang/ReductionGenClangCBTy

Similarly, elsewhere.

jdoerfert commented on clang/test/OpenMP/reduction_complex.c:
```diff
@@ -0,0 +1,1071 @@
+// NOTE: Assertions have been autogenerated by utils/update_cc_test_checks.py UTC_ARGS: --function-signature --include-generated-funcs --replace-value-regex "__omp_offloading_[0-9a-z]+_[0-9a-z]+" "reduction_size[.].+[.]" "pl_cond[.].+[.|,]" --prefix-filecheck-ir-name _
```
Add a filter here to restrict the check lines to the functions we care about.

TIFitis replied:
I've removed the checks which didn't pertain to the complex reduction section of the code.

jdoerfert reviewed (commented):
Some comments inlined.

Do we support non trivial constructors for reduction variables?
Similarly, did we support user defined reductions?
If either questions is answered with yes, do we keep that support?

We might need to ask Alexey.

TIFitis commented:
> Some comments inlined.

I've addressed these in the latest commit.

> Do we support non trivial constructors for reduction variables? Similarly, did we support user defined reductions? If either questions is answered with yes, do we keep that support?
> 
> We might need to ask Alexey.

I haven't removed any code logic or made alterations that would affect these AFAIK.
I'm not sure however if these were supported in the first place. If @alexey-bataev has any tests for them then I could try them out before and after the patch and post an update here.

alexey-bataev commented:
> I haven't removed any code logic or made alterations that would affect these AFAIK. I'm not sure however if these were supported in the first place. If @alexey-bataev has any tests for them then I could try them out before and after the patch and post an update here.

I don't remember exactly the test (added them more than 5 years ago), but both UDR and non-trivial constructors should be supported. I just remember that there were some limitations for UDR for NVPTX, like trivial UDR operations to use NVPTX fast builtins.

TIFitis commented:
> I don't remember exactly the test (added them more than 5 years ago), but both UDR and non-trivial constructors should be supported. I just remember that there were some limitations for UDR for NVPTX, like trivial UDR operations to use NVPTX fast builtins.

As long we have tests for these in check-clang or check-openmp then we should be good I guess as both are clean.

TIFitis commented:
@jsjodin @jdoerfert Hi, this patch has been sitting for a while now. Would be great to get it reviewed and accepted. Thanks.

jdoerfert commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -145,6 +145,8 @@ static bool isValidWorkshareLoopScheduleType(OMPScheduleType SchedType) {
 }
 #endif
 
+Function *GLOBAL_ReductionFunc = nullptr;
```
We should (almost) never have globals flying around. What is going on here?

TIFitis replied:
We don't need this for now, I've removed it.

jdoerfert commented on llvm/include/llvm/Frontend/OpenMP/OMPIRBuilder.h:
```diff
@@ -99,14 +100,20 @@ class OpenMPIRBuilderConfig {
 
-  // Flag for specifying if offloading is mandatory.
+  /// Flag for specifying if LLVMUsed information should be emitted.
+  std::optional<bool> EmitLLVMUsed;
```
This is a generic sounding flag that guards a very very specific thing.
Is there ever a use case where the flag is not set (and it is checked)?

TIFitis replied:
I've changed the name to `EmitLLVMUsedMetaInfo`, please let me know if you have a better name.

I don't think we have such a use case. I've updated the current code to check the flag is set before checking it.

jdoerfert commented on llvm/include/llvm/Frontend/OpenMP/OMPIRBuilder.h:
```diff
@@ -607,15 +621,17 @@ class OpenMPIRBuilder {
+  InsertPointTy createBarrier(const LocationDescription &Loc,
+                              omp::Directive Kind, bool ForceSimpleCall = false,
+                              bool CheckCancelFlag = true,
+                              Value *ThreadID = nullptr);
```
What has this to do with reductions? Either make it a pre-commit or just don't introduce this concept of "existing ThreadID values" again. Most (>90%) of the runtime functions that take a thread ID don't use it anyway. It's leftover from the old runtime.

TIFitis replied:
Most of the clang reduction tests have it in them where they reuse the ThreadID. This change allows us to pass an existing ThreadID if available.

If you'd like I could remove this change and update the tests instead, please let me know which you prefer.

jsjodin replied:
The ThreadID doesn't seem to be passed in the original code. If there was reuse before perhaps some optimization is not happening? Probably worth looking into the cause of this instead of adding the parameter.

jdoerfert replied:
This entire threadId in API stuff is only a leftover from the old (broken) runtime design.
We don't use the thread Id in barriers, and we don't need to pass it in.
If that means you need to change the tests, so be it. Do it before this commit, and feel free to modify the device RTL API to remove the argument completely.

jdoerfert commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -2096,15 +2102,1408 @@ OpenMPIRBuilder::createSection(const LocationDescription &Loc,
+    } else {
+      // LHS = Builder.CreateLoad(LHS);
+      // LHS = Builder.CreateLoad(LHS);
+      // Builder.restoreIP(RI.ReductionGen(Builder.saveIP(), LHS, RHS));
```
What is happening here?

TIFitis replied:
Code should be added here when we add MLIR/Flang codegen support. This is unreachable for now, I've replaced the commented code with an assertion failure.

jdoerfert reviewed (commented):
This patch is huge, and just skimming over it shows various places that could be split off, and others that should not make it to the review stage (commented out code).
Please update.

TIFitis commented:
> This patch is huge, and just skimming over it shows various places that could be split off, and others that should not make it to the review stage (commented out code). Please update.

Hi, thanks for talking the time out to review the patch.

I've addressed the comments you made and also fixed the build issues. It now builds cleanly again and passes all checks.

I'm not sure however on how to split the patch, as we definitely need all the pieces to pass the CI tests. I hope it isn't too much of a burden to review the patch.

Thanks.

shraiysh commented:
Not super familiar with the work here, but could we maybe split it as - one patch for OpenMPIRBuilder changes with tests in OpenMPIRBuilderTest.cpp and then second patch for the clang changes to use those OpenMPIRBuilder changes? Would the CI tests cause issues?

TIFitis commented:
> Not super familiar with the work here, but could we maybe split it as - one patch for OpenMPIRBuilder changes with tests in OpenMPIRBuilderTest.cpp and then second patch for the clang changes to use those OpenMPIRBuilder changes? Would the CI tests cause issues?

Only the functions required for OpenMP reductions directive are being moved in this patch and the OpenMPIRBuilder changes are only to facilitate this move.

No new functionality is being added.

If we were to move only a subset of the functions then we would have to add extra code to remove the dependancy on Clang CGF etc which is unnecessary.

shraiysh commented:
I see. Thanks for the reply.

TIFitis commented:
Ping for review 😄

TIFitis commented:
Ping for review.

TIFitis commented:
@jdoerfert Kind reminder for review. Thanks.

TIFitis committed:
[OpenMP] Migrate GPU Reductions CodeGen from Clang to OMPIRBuilder

This patch migrates the CGOpenMPRuntimeGPU::emitReduction and related functions to the OpenMPIRBUilder. In future patches MLIR OpenMP translation would be making use of these functions.

Co-authored-by: Jan Leyonberg <jan.leyonberg@amd.com>

jsjodin reviewed (approved):
This looks good to me now. You can wait a few days before merging to let @jdoerfert look it over again.

TIFitis committed:
Fix TargetDeInit missing TeamsReductions

dpalermo commented:
The above is an intermittent failure on the tsan buildbot (https://lab.llvm.org/buildbot/#/builders/134) and can be ignored.  The next patch processed shows status is green.

jhuber6 commented:
This patch causes the `offloading/bug51781.c` test to fail when compiled with reductions + debug information.
```console
> clang ../offload/test/offloading/bug51781.c -fopenmp -O1 --offload-arch=sm_89 -DADD_REDUCTION --offload-device-only -gline-tables-only
!dbg attachment points at wrong subprogram for function
!19 = distinct !DISubprogram(name: "__omp_offloading_10302_af88b66_main_l44", scope: !11, file: !11, line: 44, type: !20, scopeLine: 44, flags: DIFlagArtificial | DIFlagPrototyped, spFlags: DISPFlagLocalToUnit | DISPFlagDefinition | DISPFlagOptimized, unit: !10)
ptr @__omp_offloading_10302_af88b66_main_l44
  %16 = load i32, ptr %14, align 4, !dbg !50, !tbaa !35
!50 = !DILocation(line: 44, column: 58, scope: !32)
!32 = distinct !DISubprogram(name: "__omp_offloading_10302_af88b66_main_l44_omp_outlined", scope: !11, file: !11, line: 44, type: !20, scopeLine: 44, flags: DIFlagArtificial | DIFlagPrototyped, spFlags: DISPFlagLocalToUnit | DISPFlagDefinition | DISPFlagOptimized, unit: !10)
!32 = distinct !DISubprogram(name: "__omp_offloading_10302_af88b66_main_l44_omp_outlined", scope: !11, file: !11, line: 44, type: !20, scopeLine: 44, flags: DIFlagArtificial | DIFlagPrototyped, spFlags: DISPFlagLocalToUnit | DISPFlagDefinition | DISPFlagOptimized, unit: !10)
!dbg attachment points at wrong subprogram for function
!19 = distinct !DISubprogram(name: "__omp_offloading_10302_af88b66_main_l44", scope: !11, file: !11, line: 44, type: !20, scopeLine: 44, flags: DIFlagArtificial | DIFlagPrototyped, spFlags: DISPFlagLocalToUnit | DISPFlagDefinition | DISPFlagOptimized, unit: !10)
ptr @__omp_offloading_10302_af88b66_main_l44
  %14 = load i32, ptr %12, align 4, !dbg !50, !tbaa !35
!50 = !DILocation(line: 44, column: 58, scope: !32)
!32 = distinct !DISubprogram(name: "__omp_offloading_10302_af88b66_main_l44_omp_outlined", scope: !11, file: !11, line: 44, type: !20, scopeLine: 44, flags: DIFlagArtificial | DIFlagPrototyped, spFlags: DISPFlagLocalToUnit | DISPFlagDefinition | DISPFlagOptimized, unit: !10)
!32 = distinct !DISubprogram(name: "__omp_offloading_10302_af88b66_main_l44_omp_outlined", scope: !11, file: !11, line: 44, type: !20, scopeLine: 44, flags: DIFlagArtificial | DIFlagPrototyped, spFlags: DISPFlagLocalToUnit | DISPFlagDefinition | DISPFlagOptimized, unit: !10)
```

This test directly uses a reduction, and if I revert this patch it no longer breaks. I'm fairly confident that somewhere in this code we did not copy debug information correctly. Any clue where that might be?
